plugin/pkg/scheduler/core/generic_scheduler.go

```diff
@@ -422,11 +473,339 @@ func EqualPriorityMap(_ *v1.Pod, _ interface{}, nodeInfo *schedulercache.NodeInf
+				minSumPriorities = sumPriorities
+			}
+			nodeScore.sumPriorities = sumPriorities
+		}
+	}
+	if len(lowestHighPriorityNodes) == 1 {
+		return lowestHighPriorityNodes[0].nodeName
+	}
+	// There are multiple nodes with the same minimum highest priority victim.
+	// Choose the one(s) with lowest sum of priorities.
+	minNumPods := math.MaxInt32
+	lowestSumPriorityNodes := []*nodeScore{}
+	for _, nodeScore := range lowestHighPriorityNodes {
+		if nodeScore.sumPriorities == minSumPriorities {
+			lowestSumPriorityNodes = append(lowestSumPriorityNodes, nodeScore)
+			if nodeScore.numPods < minNumPods {
+				minNumPods = nodeScore.numPods
+			}
+		}
+	}
+	if len(lowestSumPriorityNodes) == 1 {
+		return lowestSumPriorityNodes[0].nodeName
+	}
+	// There are still more than one node with minimum highest priority victim and
+	// lowest sum of victim priorities. Find the anyone with minimum number of victims.
+	for _, nodeScore := range lowestSumPriorityNodes {
+		if nodeScore.numPods == minNumPods {
+			return nodeScore.nodeName
+		}
+	}
+	glog.Errorf("We should never reach here!")
+	return ""
+}
+
+// selectNodesForPreemption finds all the nodes with possible victims for
+// preemption in parallel.
+func selectNodesForPreemption(pod *v1.Pod,
+	nodeNameToInfo map[string]*schedulercache.NodeInfo,
+	nodes []*v1.Node,
+	potentialNodeNames map[string]bool,
+	predicates map[string]algorithm.FitPredicate,
+	metadataProducer algorithm.PredicateMetadataProducer,
+	extenders []algorithm.SchedulerExtender,
+) (map[string][]*v1.Pod, error) {
+
+	nodeNameToPods := map[string][]*v1.Pod{}
+	var resultLock sync.Mutex
+
+	// We can use the same metadata producer for all nodes.
+	meta := metadataProducer(pod, nodeNameToInfo)
+	checkNode := func(i int) {
+		nodeName := nodes[i].Name
+		// if potentialNodeNames is not nil, try the node only if it exists in potentialNodeNames.
+		// potentialNodeNames may be nil only in tests.
+		if potentialNodeNames != nil {
+			if _, ok := potentialNodeNames[nodeName]; !ok {
+				return
+			}
+		}
+		var metaCopy algorithm.PredicateMetadata
+		if meta != nil {
+			metaCopy = meta.ShallowCopy()
+		}
+		pods, fits := selectVictimsOnNode(pod, metaCopy, nodeNameToInfo[nodeName], predicates, extenders)
+		if fits {
+			resultLock.Lock()
+			nodeNameToPods[nodeName] = pods
+			resultLock.Unlock()
+		}
+	}
+	workqueue.Parallelize(16, len(nodes), checkNode)
+
+	// If there are any extenders, run them and filter the list of candidate nodes.
+	if len(nodeNameToPods) > 0 && len(extenders) > 0 {
+		filteredNodeNameToPods := map[string][]*v1.Pod{}
+		// Remove the victims from the corresponding nodeInfo and send nodes to the
+		// extenders for filtering.
+		for nodeName, victims := range nodeNameToPods {
+			originalNodeInfo := nodeNameToInfo[nodeName]
+			nodeInfoCopy := nodeNameToInfo[nodeName].Clone()
+			for _, victim := range victims {
+				nodeInfoCopy.RemovePod(victim)
+			}
+			nodeNameToInfo[nodeName] = nodeInfoCopy
```

Comment:
That will break stuff.

The problem happens if you return error somewhere in between, because the it won't be restored properly (and this is used on a number of places in whole scheduler).

What you should do is to opaque the body of this for loop into function and use defer to restore the nodeNameToInfo map after doing this.

Reply:
Nice catch.